src/parser/mod.rs

```diff
@@ -0,0 +1,695 @@
+		self.imported_globals_count
+	}
+
+	#[allow(dead_code)]
+	pub fn num_local_globals(&self) -> u32 {
+		self.global_types.len() as u32 - self.imported_globals_count
+	}
+
+	#[allow(dead_code)]
+	pub fn num_tags(&self) -> u32 {
+		self.tag_count
+	}
+
+	#[allow(dead_code)]
+	pub fn num_imported_tags(&self) -> u32 {
+		self.imported_tags_count
+	}
+
+	#[allow(dead_code)]
+	pub fn num_data(&self) -> u32 {
+		self.data_segments_count
+	}
+
+	#[allow(dead_code)]
+	pub fn num_elements(&self) -> u32 {
+		self.elements_count
+	}
+
+	#[allow(dead_code)]
+	pub fn num_types(&self) -> u32 {
+		self.types_map.len() as u32
+	}
+
+	#[allow(dead_code)]
+	pub fn num_export_global(&self) -> u32 {
+		self.exports_global_count
+	}
+
+	add_section_function!(Export, Export);
+	add_section_function!(Global, Global);
+	add_section_function!(Import, Import);
+	add_section_function!(Function, u32);
+	add_section_function!(Memory, MemoryType);
+	add_section_function!(Table, Table);
+	add_section_function!(Code, FunctionBody);
+	add_section_function!(Element, Element);
+}
+
+// Then insert metering calls into a sequence of instructions given the block locations and costs.
+pub fn copy_locals(func_body: &FunctionBody) -> Result<Vec<(u32, wasm_encoder::ValType)>> {
+	let mut local_reader = func_body.get_locals_reader()?;
+	// Get current locals and map to encoder types
+	let current_locals: Vec<(u32, wasm_encoder::ValType)> = (0..local_reader.get_count())
+		.map(|_| {
+			let (count, ty) = local_reader.read().unwrap();
+			(count, DefaultTranslator.translate_ty(&ty).unwrap())
+		})
+		.collect::<Vec<(u32, wasm_encoder::ValType)>>();
+
+	Ok(current_locals)
+}
+
+// TODO unable to get function encoder body directly, remove this after option wasmparser
+pub fn truncate_len_from_encoder(func_builder: &dyn wasm_encoder::Encode) -> Result<Vec<u8>> {
+	let mut d = vec![];
+	func_builder.encode(&mut d);
+	let mut r = wasmparser::BinaryReader::new(&d);
+	let size = r.read_var_u32()?;
+	Ok(r.read_bytes(size as usize)?.to_vec())
+}
+
+fn rebuild_name_map(name_map: NameMap) -> Result<wasm_encoder::NameMap> {
+	let mut encoded_map = wasm_encoder::NameMap::new();
+	for naming in name_map {
+		let naming = naming?;
+		encoded_map.append(naming.index, naming.name);
+	}
+	Ok(encoded_map)
+}
+
+fn rebuild_indirect_name_map(
+	indirect_name_map: IndirectNameMap,
+) -> Result<wasm_encoder::IndirectNameMap> {
+	let mut encoded_map = wasm_encoder::IndirectNameMap::new();
+	for indirect_naming in indirect_name_map {
+		let indirect_naming = indirect_naming?;
+		let new_map = rebuild_name_map(indirect_naming.names)?;
+
+		encoded_map.append(indirect_naming.index, &new_map);
+	}
+	Ok(encoded_map)
+}
+
+/// - Update function indices in Custom Name section
+///   Increment indices greater or equal than given one (incrementing by 1 because it is assumed one
+///   function has been added before the given index)
+/// - Rebuild remaining section items
+pub fn process_custom_section(
```

Comment:
Can move `rebuild_name_map`,  `rebuild_indirect_name_map` and `process_custom_section` outside `parser` mod? Not reviewed either.

Reply:
`parser` module renamed to util (as in fvm-wasm-instrument)
and it consist of:
```
pub mod module_info;
pub mod transform;  <- moved here: rebuild_name_map, rebuild_indirect_name_map and process_custom_section
pub mod translator;
```